flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/flip1/ExecutionFailureHandlerTest.java

```diff
@@ -0,0 +1,170 @@
+		when(failoverStrategy.getTasksNeedingRestart(any(ExecutionVertexID.class), any(Throwable.class))).thenReturn(
+			tasksToBeRestarted);
+
+		long restartDelayMs = 1234;
+		when(restartStrategy.getBackoffTime()).thenReturn(restartDelayMs);
+
+		// suppress restarts
+		when(restartStrategy.canRestart()).thenReturn(false);
+
+		FailureHandlingResult result = executionFailureHandler.getFailureHandlingResult(
+			new ExecutionVertexID(new JobVertexID(), 0),
+			new Exception("test failure"));
+
+		// verify results
+		assertFalse(result.canRestart());
+		assertNotNull(result.getError());
+		try {
+			result.getVerticesToRestart();
+			fail("get tasks to restart is not allowed when restarting is suppressed");
+		} catch (IllegalStateException ex) {
+			// expected
+		}
+		try {
+			result.getRestartDelayMS();
+			fail("get restart delay is not allowed when restarting is suppressed");
+		} catch (IllegalStateException ex) {
+			// expected
+		}
+	}
+
+	/**
+	 * Tests the case that the failure is non-recoverable type.
+	 */
+	@Test
+	public void testNonRecoverableFailureHandlingResult() throws Exception {
+		ExecutionFailureHandler executionFailureHandler = createExecutionFailureHandler();
+		FailoverStrategy failoverStrategy = executionFailureHandler.getFailoverStrategy();
+		RestartBackoffTimeStrategy restartStrategy = executionFailureHandler.getRestartBackoffTimeStrategy();
+
+		Set<ExecutionVertexID> tasksToBeRestarted = new HashSet<>();
+		when(failoverStrategy.getTasksNeedingRestart(any(ExecutionVertexID.class), any(Throwable.class))).thenReturn(
+			tasksToBeRestarted);
+
+		long restartDelayMs = 1234;
+		when(restartStrategy.getBackoffTime()).thenReturn(restartDelayMs);
+
+		// accept restarts
+		when(restartStrategy.canRestart()).thenReturn(true);
+
+		FailureHandlingResult result = executionFailureHandler.getFailureHandlingResult(
+			new ExecutionVertexID(new JobVertexID(), 0),
+			new SuppressRestartsException(new Exception("test failure")));
+
+		// verify results
+		assertFalse(result.canRestart());
+		assertNotNull(result.getError());
+		try {
+			result.getVerticesToRestart();
+			fail("get tasks to restart is not allowed when restarting is suppressed");
+		} catch (IllegalStateException ex) {
+			// expected
+		}
+		try {
+			result.getRestartDelayMS();
+			fail("get restart delay is not allowed when restarting is suppressed");
+		} catch (IllegalStateException ex) {
+			// expected
+		}
+	}
+
+	// ------------------------------------------------------------------------
+	//  utilities
+	// ------------------------------------------------------------------------
+
+	private ExecutionFailureHandler createExecutionFailureHandler() {
+		FailoverTopology topology = new TestFailoverTopology.Builder().build();
+
+		FailoverStrategy.Factory failoverStrategyFactory = mock(FailoverStrategy.Factory.class);
+		when(failoverStrategyFactory.create(any(FailoverTopology.class))).thenReturn(mock(FailoverStrategy.class));
```

Comment:
Please create a proper implementation of the FailoverStrategy. The factory can then be a simple lambda `topology -> new MyFailoverStrategy`.

Mocking things usually just leads to problems down the line.

Reply:
Ok.